### DISCUSSION TITLE
Race condition found

### DISCUSSION THREAD

**Discussion by erezrob** [psollberger/DeriSock#30] [General]
I found a bug, I hope someone can fix and integrate this into the library in the future.

When subscribing to a book (for example: SendAsync: {"Original":null,"JsonRpc":"2.0","Id":19,"Method":"public/subscribe","Params":{"channels":["book.BTC-17JUN22-32000-P.100ms"]},"$type":"JsonRpcRequest"}

What happens is, that there are 2 response messages that follow (not one as in most cases), one is the normal "I got your sub request" and the second one is "Snapshot", this gives us the current book starting point of the asset before we start getting the stream of changes. The problem is that because "InternalOnResponseReceived" is handeled async, we sometimes get the first response and then go to: Subscribe: Successfully subscribed. Adding callback
And sometimes we only add the callback after both responses are received.

The problem is that if the callback is constructed after the snapshot is received we will never ever bubble to the user code that a snapshot has been received.

Here is an example of the bug in log:
  (186):[12:08:22 VRB] SendAsync: {"Original":null,"JsonRpc":"2.0","Id":19,"Method":"public/subscribe","Params":{"channels":["book.BTC-17JUN22-32000-P.100ms"]},"$type":"JsonRpcRequest"}
  (226):[12:08:22 VRB] ProcessReceive: Received Message (145 ; 0.005ms): {"jsonrpc":"2.0","id":19,"result":["book.BTC-17JUN22-32000-P.100ms"],"usIn":1655024902246846,"usOut":1655024902246895,"usDiff":49,"testnet":true}
  (228):[12:08:22 VRB] ProcessReceive: Received Message (250 ; 0.003ms): {"jsonrpc":"2.0","method":"subscription","params":{"channel":"book.BTC-17JUN22-32000-P.100ms","data":{"type":"snapshot","timestamp":1655024897743,"instrument_name":"BTC-17JUN22-32000-P","change_id":12337361264,"bids":[["new",0.0735,2.9]],"asks":[]}}}
  (231):[12:08:22 VRB] OnNotification: {"Timestamp":"0001-01-01T00:00:00.0000000","Params":{"Channel":"book.BTC-17JUN22-32000-P.100ms","Data":[[[]],[[]],[[]],[[]],[[[[],[],[]]]],[[]]],"$type":"NotificationParams"},"Channel":"book.BTC-17JUN22-32000-P.100ms","Data":[[[]],[[]],[[]],[[]],[[[[],[],[]]]],[[]]],"Original":null,"JsonRpc":"2.0","Id":0,"Method":"subscription","$type":"Notification"}
  (272):[12:08:22 DBG] Subscribe: Successfully subscribed. Adding callback (Channel: book.BTC-17JUN22-32000-P.100ms)
  (277):[12:08:22 VRB] SendAsync: {"Original":null,"JsonRpc":"2.0","Id":41,"Method":"public/subscribe","Params":{"channels":["trades.BTC-17JUN22-32000-P.100ms"]},"$type":"JsonRpcRequest"}
  (352):[12:08:22 VRB] ProcessReceive: Received Message (147 ; 0.006ms): {"jsonrpc":"2.0","id":41,"result":["trades.BTC-17JUN22-32000-P.100ms"],"usIn":1655024902372541,"usOut":1655024902372598,"usDiff":57,"testnet":true}
  (368):[12:08:22 DBG] Subscribe: Successfully subscribed. Adding callback (Channel: trades.BTC-17JUN22-32000-P.100ms)

**Comment by psollberger** [comment]
Hi @erezrob 

Thank you for bringing this to my attention. I'll investigate on that.

**Comment by badokun** [comment]
On a related note, in a separate project I've implemented some integration tests for this library. That would make debugging and fixing this type of issue much easier imo. Let me know if you're interested

**Reply by psollberger** [reply]
I have an open issue #21 for that. If you want to contribute on that, i'm more than happy to merge that.
